storage/replica_command.go

```diff
@@ -756,10 +756,8 @@ func (r *Replica) PushTxn(batch engine.Engine, ms *engine.MVCCStats, args proto.
 		}
 	} else {
 		// Some sanity checks for case where we don't find a transaction record.
-		if args.PusheeTxn.LastHeartbeat != nil {
-			return reply, proto.NewTransactionStatusError(&args.PusheeTxn,
-				"no txn persisted, yet intent has heartbeat")
-		} else if args.PusheeTxn.Status != proto.PENDING {
+		// If `args.PusheeTxn.LastHeartbeat != nil`, assume txn is gc'ed, see #2111.
```

Comment:
If we write a PENDING entry, it could mean having pushed the Txn into the future. That means that it's going to update the intent (see `MVCCResolveWriteIntent`). I've checked that it'll be a no-op since the epoch hasn't increased:
```go
	// This method shouldn't be called in this instance, but there's
	// nothing to do if meta's epoch is greater than or equal txn's
	// epoch and the state is still PENDING.
```

so it ends there (and nobody's going to commit this resuscitated transaction anyways), but we still end up keeping the intent and having a fresh and shiny transaction table entry for others to have to play the conflict resolution game with.

Discussing that with the above seems adequate so that we avoid recreating the entry and allow the push when we know that it's GC at work (being able to do this depends on #2062 though).